### ISSUE TITLE
Update tests that have the wrong includes

### ISSUE THREAD

**Issue by syclik** [stan-dev/math#251]
#### Summary:
I missed some tests when I fixed #236.

#### Description:
Some of the tests include more headers than just the one it needs.

#### Current Version:
v2.9.0

**Comment by syclik** [comment]
@bgoodri or @thelseraphim, do you know how we could run `iwyu` on header files and get meaningful output?
https://github.com/stan-dev/stan/wiki/include-what-you-use

We don't want to run it on the big include files: `stan/math/{prim, rev, fwd, mix}/{scal, arr, mat}.hpp`, but we want to run it on all the include files past that. I'm not even sure how to call it on one file.

**Comment by bgoodri** [comment]
I don't know how to prevent iwyu from running on meta-headers, but you can just ignore the output for those files. It often gets confused by Eigen.

`find stan/math/ -name '*.hpp' -print0 | parallel -0 -j8 make -k -j1 {}-test CC=iwyu "&>" {}_iwyu.out`
`find . -name '*_iwyu.out' -print0 | xargs -0 tar -rf /tmp/iwyu.tar`
`gzip /tmp/iwyu.tar`
`find stan/math -name "*_iwyu.out" -type f -print0 | xargs -0 /bin/rm -f`

**Comment by peterwicksstringfield** [comment]
I figured out how to use iwyu on our header files automatically, if there is still interest?

Take a look at this output:
```
peter@gtower:~/stan/math$ make test-iwyu
./runiwyu.sh -Xiwyu --check_also='*hpp' -Xiwyu --no_fwd_decls -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare      -I lib/tbb_2019_U8/include -O3  -I . -I lib/eigen_3.3.3 -I lib/boost_1.72.0 -I lib/sundials_5.1.0/include -O0 stan/math/mix/fun/typedefs.hpp 2>&1 | tee tmp_iwyu_output | grep -q "has correct #includes/fwd-decls" || cat tmp_iwyu_output
./runiwyu.sh -Xiwyu --check_also='*hpp' -Xiwyu --no_fwd_decls -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare      -I lib/tbb_2019_U8/include -O3  -I . -I lib/eigen_3.3.3 -I lib/boost_1.72.0 -I lib/sundials_5.1.0/include -O0 stan/math/mix/functor/grad_hessian.hpp 2>&1 | tee tmp_iwyu_output | grep -q "has correct #includes/fwd-decls" || cat tmp_iwyu_output

stan/math/mix/functor/grad_hessian.hpp should add these lines:
#include <ext/alloc_traits.h>                          // for __alloc_traits...
#include "Core"                                        // for Matrix, DenseC...
#include "stan/math/fwd/core/fvar.hpp"                 // for fvar
#include "stan/math/rev/core/grad.hpp"                 // for grad
#include "stan/math/rev/core/nested_rev_autodiff.hpp"  // for nested_rev_aut...
#include "stan/math/rev/core/var.hpp"                  // for var

stan/math/mix/functor/grad_hessian.hpp should remove these lines:
- #include <stan/math/fwd/core.hpp>  // lines 4-4
- #include <stan/math/prim/fun/Eigen.hpp>  // lines 5-5
- #include <stan/math/rev/core.hpp>  // lines 6-6
- #include <stdexcept>  // lines 7-7

The full include-list for stan/math/mix/functor/grad_hessian.hpp:
#include <ext/alloc_traits.h>                          // for __alloc_traits...
#include <vector>                                      // for vector
#include "Core"                                        // for Matrix, DenseC...
#include "stan/math/fwd/core/fvar.hpp"                 // for fvar
#include "stan/math/rev/core/grad.hpp"                 // for grad
#include "stan/math/rev/core/nested_rev_autodiff.hpp"  // for nested_rev_aut...
#include "stan/math/rev/core/var.hpp"                  // for var
---
... [lots more output]
```

Here I tweaked mix/fun/typedefs.hpp so I could give an example of a file that iwyu wouldn't flag.

You can see it complaining about stuff in mix/functor/grad_hessian.hpp.

It wants use to add "stan/math/fwd/core/fvar.hpp", but <stan/math/fwd/core/fvar.hpp> works too, so we don't need to switch to relative-includes to keep iwyu happy or anything like that.

There are some issues with Eigen, but they might be managable.

I added this target to make/tests:
```
############################################################
#
# Target to verify header files within Stan include
# as little as possible
##
IWYU_TESTS := $(addsuffix -iwyu,$(call findfiles,stan,*.hpp))

%.hpp-iwyu : %.hpp
	./runiwyu.sh -Xiwyu --check_also='*hpp' -Xiwyu --no_fwd_decls $(CXXFLAGS) -O0 $^ 2>&1 | tee tmp_iwyu_output | grep -q "has correct #includes/fwd-decls" || cat tmp_iwyu_output

.PHONY: test-iwyu
test-iwyu: $(IWYU_TESTS)
```

That's not production quality, but it demonstrates the idea I think.

Basically "make test-iwyu" delegates to "make stan/math/prim/err/check_nonnegative.hpp-iwyu", and the compile command for that is to invoke iwyu and check the output. If the output includes "has correct #includes/fwd-decls" then we throw away the output. If the output does not include that phrase we print it out.

"--check_also="*hpp"" because iwyu by default only gives output for things that look like cpp files.
"--no_fwd_decls" because by default iwyu prefers forward declaring things to include the headers for them, and we prefer to include headers.

runiwyu.sh is just:
```
#!/bin/bash

../include-what-you-use/build/include-what-you-use $@
```

I had to compile iwyu locally because the version in my distro's repositories was too old to support the --no_fwd_decls argument.

**Comment by bob-carpenter** [comment]
Thanks for looking into this.

I think some of these might be dangerous because of order-of-include issues.  That's not being preserved, is it?

The bigger issue is which of those headers is better for understandability and longer-term maintainability.  This is directly related to the level of includes.  It's much much easier and more maintainable to just include the whole of `rev/core.hpp` than to cherry pick exactly what's needed.  For example, the suggested replacement includes two header files, `"Core"` and `<ext/alloc_traits.h>`, that I've never heard of.  And `nested_rev_autodiff.hpp` is relatively obscure as a direct include.

> iwyu prefers forward declaring things over including the headers for them, and we prefer to include headers.

Any idea why?

**Comment by SteveBronder** [comment]
Looks neat! But yeah along with Bob my concern is that it's kind of toilsome to try to get exactly the right headers need for each header. I remember working on the gp_exp_quad stuff and spending like 2-3 hours banging my head on the keyboard trying to get the includes right. including a top level header instead makes things a lot easier

If you are interested in adding some tooling to stan I think it would be v cool to have easy access to fuzzers, static analysis tools, and sanitizers.

**Comment by peterwicksstringfield** [comment]
>Any idea why?

Something about an old version of the google style guide. I didn't really look into it.

The "Core" header is Eigen/Core ... I don't know why its just called Core. That can actually be fixed with a thing iwyu calls an include mapping. alloc_traits is a new one on me as well.

I guess I should clarify I'm not suggesting anything. As far as I'm concerned, now that their are kinda-sorta-working instructions for how to use iwyu attached to this ticket, that's mission accomplished.

If we wanted to actually use this ... it would be a mistake to try to add it to the build process. If anything someone might want to use it to look for stuff and then fix it manually. So for example, plain_type includes itself; and apply_vector_unary uses is_container but doesn't (directly) include it. I'm not even suggesting that.

**Comment by bob-carpenter** [comment]
Even the header tests can be challenging, because of the prim vs. autodiff distinction.  I wound up having to include more than I'd like there for some functions.  That's sort of the opposite problem, though.
